pkg/cloud/services/networking/port.go

```diff
@@ -393,112 +333,155 @@ func (s *Service) CreatePorts(eventObject runtime.Object, clusterName, baseName
-			if port.Trunk != nil && *port.Trunk {
+		for _, port := range resolvedPorts {
+			if port.Trunk {
 				return true
 			}
 		}
 		return false
 	}
 	if portUsesTrunk() {
+		trunkSupported, err := s.IsTrunkExtSupported()
+		if err != nil {
+			return nil, err
+		}
+
 		if !trunkSupported {
 			return nil, fmt.Errorf("there is no trunk support. please ensure that the trunk extension is enabled in your OpenStack deployment")
 		}
 	}
 
-	return ports, nil
+	return resolvedPorts, nil
 }
 
 // normalizePorts ensures that a user-specified PortOpts has all required fields set. Specifically it:
 // - sets the Trunk field to the instance spec default if not specified
 // - sets the Network ID field if not specified.
-func (s *Service) normalizePorts(ports []infrav1.PortOpts, openStackCluster *infrav1.OpenStackCluster, trunkEnabled bool) ([]infrav1.PortOpts, error) {
-	normalizedPorts := make([]infrav1.PortOpts, 0, len(ports))
+func (s *Service) normalizePorts(ports []infrav1.PortOpts, clusterName, baseName string, trunkEnabled bool, defaultSecurityGroupIDs []string, defaultNetwork *infrav1.NetworkStatusWithSubnets, baseTags []string) ([]infrav1.ResolvedPortSpec, error) {
+	normalizedPorts := make([]infrav1.ResolvedPortSpec, len(ports))
 	for i := range ports {
-		// Deep copy the port to avoid mutating the original
-		port := ports[i].DeepCopy()
+		port := &ports[i]
+		normalizedPort := &normalizedPorts[i]
+
+		// Copy fields which don't need to be resolved
+		normalizedPort.ResolvedPortSpecFields = port.ResolvedPortSpecFields
+
+		// Generate a standardised name
+		normalizedPort.Name = getPortName(baseName, port, i)
+
+		// Generate a description if none is provided
+		if port.Description != nil {
+			normalizedPort.Description = *port.Description
+		} else {
+			normalizedPort.Description = names.GetDescription(clusterName)
+		}
+
+		// Tags are inherited base tags plus any port-specific tags
+		normalizedPort.Tags = slices.Concat(baseTags, port.Tags)
 
 		// No Trunk field specified for the port, inherit the machine default
 		if port.Trunk == nil {
-			port.Trunk = &trunkEnabled
+			normalizedPort.Trunk = trunkEnabled
+		} else {
+			normalizedPort.Trunk = *port.Trunk
 		}
 
-		if err := s.normalizePortTarget(port, openStackCluster, i); err != nil {
+		// Resolve network ID and fixed IPs
+		var err error
+		normalizedPort.NetworkID, normalizedPort.FixedIPs, err = s.normalizePortTarget(port, defaultNetwork, i)
+		if err != nil {
 			return nil, err
 		}
 
-		normalizedPorts = append(normalizedPorts, *port)
+		// Resolve security groups
+		if len(port.SecurityGroups) == 0 {
+			normalizedPort.SecurityGroups = defaultSecurityGroupIDs
+		} else {
+			normalizedPort.SecurityGroups, err = s.GetSecurityGroups(port.SecurityGroups)
+			if err != nil {
+				return nil, fmt.Errorf("error getting security groups: %v", err)
+			}
+		}
 	}
 	return normalizedPorts, nil
 }
 
-// normalizePortTarget ensures that the port has a network ID.
```

Comment:
I think all below should be double-checked by @MaysaMacedo. It looks right to me from dual stack perspective, but this is the delicate part in this PR.

Reply:
It should be just code motion with changes for context. I didn't intentionally change the logic here.